### PR TITLE
[core] create bntable

### LINKED ISSUE #752: [core] Create BnTable

### PR DESCRIPTION
**Pull request by iurynogueira**
## Issue Number

fix #752

## Description

Create a implementation of smart table

### CODE REVIEW

File: projects/ion/src/lib/use-table/use-table.component.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  selector: 'ion-use-table',
+  templateUrl: './use-table.component.html',
+  styleUrls: ['./use-table.component.scss'],
+})
+export class IonUseTableComponent extends BnTable<User> {
+  constructor(service: IonApiService) {
+    super({
+      service,
+      tableConfig: {
+        columns: [{ label: 'Nome', sort: true, key: 'nome' }],
```
Comment (MarcosOliveir4) [inline projects/ion/src/lib/use-table/use-table.component.ts:21]: Will you ever need to have a name column?

**Review by MarcosOliveir4 — CHANGES_REQUESTED** [review]

File: projects/ion/src/core/api/http.interfaces.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+export interface BnService<DataType> {
+  list: (filters?: IPayload) => Observable<IResponse<DataType>>;
+  insert?: (payload: IPayload) => Observable<SafeAny>;
+}
+
+export interface IResponse<DataType> {
+  data?: DataType[];
+  dados?: DataType[];
+  total?: number;
+  mensagem?: string;
```
[the remark this answers is not in the document]
Comment (iurynogueira) [reply projects/ion/src/core/api/http.interfaces.ts:14]: > On some systems is msg returned yet?

Yep!

File: projects/ion/src/core/bn-table/bn-table.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    const dataRequest$ = this.service
+      .list({ ...this.payload, total: false })
+      .pipe(take(1));
+
+    forkJoin([totalRequest$, dataRequest$]).subscribe(
+      (response) => {
+        this.configTable.loading = false;
+
+        const totalResponse: IResponse<DataType> = response[0];
```
[the remark this answers is not in the document]
Comment (iurynogueira) [reply projects/ion/src/core/bn-table/bn-table.ts:86]: > I believe there is a way to unstructure the return

Good idea, response not is a object, is a array but we can do const [total, data] = response. What do you think?

File: projects/ion/src/core/utils/clearObject.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { SafeAny } from '../../lib/utils/safe-any';
+import { isEmpty } from 'lodash';
```
[the remark this answers is not in the document]
Comment (iurynogueira) [reply projects/ion/src/core/utils/clearObject.ts:2]: > Is it really necessary to use lodash?

We can remove and use other way.

[the remark this answers is not in the document]
Comment (iurynogueira) [reply projects/ion/src/core/utils/clearObject.ts:2]: @jAllanOli

File: projects/ion/src/lib/use-table/use-table.component.ts (hunk 1 above)
Comment (iurynogueira) [reply projects/ion/src/lib/use-table/use-table.component.ts:21]: > Will you ever need to have a name column?

This component its for tests on storybook.

File: projects/ion/src/core/bn-table/bn-table.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (iurynogueira) [reply projects/ion/src/core/bn-table/bn-table.ts:86]: changed.

File: projects/ion/src/core/utils/clearObject.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jAllanOli) [reply projects/ion/src/core/utils/clearObject.ts:2]: Got it

File: projects/ion/src/core/utils/clearObject.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+function isNestedObj(value: SafeAny): boolean {
+  return value && !Array.isArray(value) && typeof value === 'object';
+}
+
+function shouldDelete(value: SafeAny): boolean {
+  return (
+    [null, undefined, ''].includes(value) ||
+    Number.isNaN(value) ||
+    (Array.isArray(value) && isEmpty(value))
```
[the remark this answers is not in the document]
Comment (deeved-hiuston-brisa) [reply projects/ion/src/core/utils/clearObject.ts:25]: Se a função resolver, acredito que seja uma boa ideia, já que não dependemos de loadsh.

**Review by deeved-hiuston-brisa — APPROVED** [review]

File: projects/ion/src/core/utils/clearObject.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (iurynogueira) [reply projects/ion/src/core/utils/clearObject.ts:25]: > What do you think about changing isEmpty for this function? `function isEmpty(value) { if (value === null) { return true; } if (typeof value === 'string' || Array.isArray(value)) { return value.length === 0; } if (typeof value === 'object') { for (const key in value) { if (Object.hasOwnProperty.call(value, key)) { return false; } } return true; } return false; } `

Thanks for this code :)

**Review by MarcosOliveir4 — APPROVED** [review]

**Review by jAllanOli — APPROVED** [review]

**Pull request merged**